internal/tofu/transform_provider.go

```diff
@@ -108,61 +118,78 @@ func (t *ProviderTransformer) Transform(g *Graph) error {
 	// Our "requested" map is from graph vertices to string representations of
 	// provider config addresses (for deduping) to requests.
 	type ProviderRequest struct {
-		Addr  addrs.AbsProviderConfig
-		Exact bool // If true, inheritance from parent modules is not attempted
+		Addr        addrs.AbsProviderConfig
+		Exact       ExactProvider // If populated, inheritance from parent modules is not attempted
+		instanceKey addrs.InstanceKey
 	}
 	requested := map[dag.Vertex]map[string]ProviderRequest{}
 	needConfigured := map[string]addrs.AbsProviderConfig{}
 	for _, v := range g.Vertices() {
```

Comment:
can we reduce nesting inside of this loop?

Reply:
Probably. I think we should refactor this code eventually. But because it is a little delicate and quite critical, I believe we should first release it to alpha and then refactor.